### PR TITLE
set phpstan to bleedingEdge config

### PR DESCRIPTION
**Pull request by Flyingmana**
there is a performance improvement for phpstan possible when switching to the currend bleedingEdge config.

https://phpstan.org/blog/phpstan-1-6-0-with-conditional-return-types#lower-memory-consumption

Heard about it while researching for a different project ( https://github.com/phpstan/phpstan/discussions/8813 )

### Questions or comments

There are some incompatibilities mentioned, and while I checked I also saw that we get some errors.
Not sure yet about what needs to change, but for the future we will need to fix it anyway anywhen

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All automated tests passed successfully (all builds are green)
 - [ ] Add yourself to contributors list

### CODE REVIEW

**Comment by elidrissidev** [comment]
Nice, I noticed an improvement from 1:35 to only 40s when using the new configuration!

**Comment by elidrissidev** [comment]
Fixed fix for some phpstan issues, but some tricky ones popped up.

File: index.php [hunk 1]
```diff
@@ -28,6 +28,7 @@
 require MAGENTO_ROOT . '/app/bootstrap.php';
 require_once $mageFilename;
 
+
```
Comment (fballiano) [inline index.php:31]:
```suggestion
```

File: app/code/core/Mage/Adminhtml/Block/Catalog/Category/Abstract.php [hunk 2]
```diff
@@ -40,8 +40,9 @@ public function getCategory()
 
     public function getCategoryId()
     {
-        if ($this->getCategory()) {
-            return $this->getCategory()->getId();
+        $category = $this->getCategory();
+        if (is_object($category) && $category->getId()) {
```
Comment (fballiano) [inline app/code/core/Mage/Adminhtml/Block/Catalog/Category/Abstract.php:44]:
```suggestion
        if (is_object($category)) {
```

@elidrissidev what about this?

**Comment by fballiano** [comment]
fixed conflict with baseline by using 1.9.4.x baseline

**Comment by fballiano** [comment]
guys, this is so incredibly much faster that I suggest we regenerate the baseline based on the code that's now in the PR and merge it. what do you think?

**Comment by fballiano** [comment]
I've regenerated the baseline (from a linux machine, to avoid problems with macos), we could merge it probably

File: app/code/core/Mage/Adminhtml/Block/Catalog/Category/Abstract.php (hunk 2 above)
Comment (elidrissidev) [reply app/code/core/Mage/Adminhtml/Block/Catalog/Category/Abstract.php:44]: Actually now that I think about it, what was reason of changing the `if ($this->getCategory())` condition? The only PHPStan error in this file was about the `Right side of && is always false` which you already fixed below..

**Comment by elidrissidev** [comment]
Resolved some issues that were newly added to baseline since they were easy to fix.

File: app/code/core/Mage/Adminhtml/Block/Catalog/Category/Abstract.php [hunk 3]
```diff
@@ -53,7 +54,8 @@ public function getCategoryName()
 
     public function getCategoryPath()
     {
-        if ($this->getCategory()) {
+        $category = $this->getCategory();
+        if (is_object($category) && $category->getId()) {
```
Comment (fballiano) [inline app/code/core/Mage/Adminhtml/Block/Catalog/Category/Abstract.php:58]:
```suggestion
        if ($this->getCategory()) {
```

File: app/code/core/Mage/Adminhtml/Block/Catalog/Category/Abstract.php (hunk 2 above)
Comment (fballiano) [inline app/code/core/Mage/Adminhtml/Block/Catalog/Category/Abstract.php:44]:
```suggestion
        if ($this->getCategory()) {
```

**Review by fballiano — APPROVED** [review]

File: app/code/core/Mage/Adminhtml/Block/Catalog/Category/Abstract.php [hunk 4]
```diff
@@ -93,8 +93,9 @@ public function getRoot($parentNodeCategory = null, $recursionLevel = 3)
             $tree = Mage::getResourceSingleton('catalog/category_tree')
                 ->load(null, $recursionLevel);
 
-            if ($this->getCategory()) {
-                $tree->loadEnsuredNodes($this->getCategory(), $tree->getNodeById($rootId));
+            $category = $this->getCategory();
+            if (is_object($category) and $category->getId()) {
```
Comment (justinbeaty) [inline app/code/core/Mage/Adminhtml/Block/Catalog/Category/Abstract.php:97]: Not a big deal, but using `&&` would be more consistent with the codebase.

```php
if (is_object($category) && $category->getId()) {
```

Comment (fballiano) [inline app/code/core/Mage/Adminhtml/Block/Catalog/Category/Abstract.php:97]:
```suggestion
            if (is_object($category) && $category->getId()) {
```

Comment (fballiano) [reply app/code/core/Mage/Adminhtml/Block/Catalog/Category/Abstract.php]: true, done

Comment (elidrissidev) [reply app/code/core/Mage/Adminhtml/Block/Catalog/Category/Abstract.php]: what is this change for?

**Comment by fballiano** [comment]
@elidrissidev  I lost your last comment, I guess that change was an older one (since this PR went thru many revisions), I've removed it, let's see if it works

**Review by elidrissidev — APPROVED** [review]
Looks good

**Review by fballiano — APPROVED** [review]

**Comment by fballiano** [comment]
merged and v20ed

**Comment by elidrissidev** [comment]
btw when you clear the description of the final squashed commit, be sure to keep the any special lines in it such as `Signed-off-by`,`Co-authored-by`, etc. Else that information will not appear on the merged commit

**Comment by fballiano** [comment]
sorry, sometimes there's so much stuff in there that doesn't seem useful at all :-(

**Comment by elidrissidev** [comment]
No problem :)

**Pull request merged**